### ISSUE TITLE
Fail fast when running parallel tasks

### ISSUE THREAD

**Issue by DavidPerezIngeniero** [sbt/sbt#2450]
Given this simple `build.sbt`:

``` scala
val a = taskKey[Unit]("A")
a := {
  Thread.sleep(500)
  sys error "A big error"
}
val b = taskKey[Unit]("B")
b := {
  Thread.sleep(5000)
  println("b")
}
val c = taskKey[Unit]("C")
c := {
  a.value
  b.value
  println("c")
}
```

I expect that when I run `c` task, to fail fast when `a` fails and not to wait the completion of long running tasks like `b`.  Before failing, it waits for the completion of `b`.

**Comment by eed3si9n** [comment]
Are you suggesting that we interrupt the `b.value`'s task execution somehow?
How would this work with task aggregation across the subprojects? Should a test or compilation failure stop the execution of all tests/compilation?

**Comment by DavidPerezIngeniero** [comment]
Yes, or at least to have the chance to do so.  Sometimes if something fails, it doesn't make sense to wait for the completion of other tasks.

Instead of `Def.sequence()`, we could have something like `Def.parallel()` with this behavior (fail fast).
And let the default behavior of waiting all taks, even if some of the fail by using:

```
a.value
b.value
```

**Comment by eed3si9n** [comment]
A related feature in sbt is

``` scala
cancelable in Global := true
```

which I don't know how well it works in practice. Do you think `Ctrl+C` could work for your use case?

**Comment by SethTisue** [comment]
I'm far from convinced this would be desirable default behavior. b might have run a long time already and be almost ready to produce a successful result. canceling it throws away the work it did.

**Comment by eed3si9n** [comment]
@SethTisue So let's say `a` and `b` represent subproject publishing. One can argue that it's desirable to cancel `b`'s side effect (_cough_, publishing) as soon as we find out that the overall execution is going to be a bad one.

**Comment by DavidPerezIngeniero** [comment]
Thanks Eugene, I didn't know about `cancelable`.

@eed3si9n , that's the point if a test fails, I'd like to cancel building a installer and deploy to a test server.  I don't care about unfinished work, but to be quick.

I'm not talking about changing default behavior, but to be able to choose the behavior.  There are cases where default behavior is ok.

**Comment by eed3si9n** [comment]
Could you try with `cancelable` aka `Ctrl+C` for a while, and see how well it works for you?
I've heard bad things like Ivy cache corruption if you just forcefully kill sbt process in the middle of a task.

**Comment by DavidPerezIngeniero** [comment]
Ctrl+C is not so useful when used in a Jenkins.

But `cancelable` is very useful for interactive use.  Sometimes I mistype a long running command, and being able to cancel it, is good news for me, but this use case isn't related with the fail fast isssue.

**Comment by agilesteel** [comment]
Just stumbled on this one. So is there a way to do this on CI? I created a classic task to format, run tests, build docker image and I want them to run concurrently so that compilation does not need to wait for the formatting checks to finish or the docker image build does not need to wait for the tests to pass (I simply won't publish the image if things go wrong, but I don't mind building it).

It works great, but I'd also want a formatting check failure to abort the very long compilation/test processes.

**Comment by agilesteel** [comment]
I guess not...

**Comment by dwalend** [comment]
I just want to stop and see any test failure before it scrolls off-screen.
